### DISCUSSION TITLE
An option (or workaround) that ensures that step previews are always detected as Regex

### DISCUSSION THREAD

**Discussion by RikvanSpreuwel** [reqnroll/Reqnroll#351] [Ideas]
Would it be possible to globally ensure that all step definitions are always detected as regex for the preview in feature files? (without adding '$^' to every step definition).

Note: the forceRegexPlugin workaround described here does not fix this issue https://docs.reqnroll.net/latest/guides/how-to-configure-cucumber-expression-behavior.html

With preview I mean that when you type _'When'_ or _'Then'_ etc on a new line, you see all possible steps filtered, which are filtered based on what you type next. E.g. 'When button' presents this list of steps as possibilities in our project: 
![image](https://github.com/user-attachments/assets/df87cbdd-c8c0-4205-bc77-fe522aadfba5)

Several steps here are not correctly recognized as regex because they do not use certain patterns. An example of this is the step **"the user clicks on '[string]' button for [string] '[string]'"**, which should be **"the user clicks on '([^']*)' button for (folder|program) '([^']*)'"**. 
See step definition: 
![image](https://github.com/user-attachments/assets/6c24e4f0-9621-4c7b-a3bb-fb60d4d896e4)

The main issue with that step definitions are not recognized as regex in this case are as follows:

- specific string options are shown as **[string]** rather than e.g. **(folder|program)**, as a result, one has to guess what string belongs here
- if you would search for a step _When program_, the example step will not be shown as an option. This results in steps being difficult to find and sometimes being unnecessarily duplicated

**Comment by gasparnagy** [comment]
Yes, this might help: https://docs.reqnroll.net/latest/guides/how-to-configure-cucumber-expression-behavior.html

**Reply by RikvanSpreuwel** [reply]
Hi, thank you for the response. However, as I mentioned, the ForceRegexPlugin doesnt work for the step preview. 

It does make sure that steps are correctly seen as regex for running tests, but we want to make sure that the steps are correctly detected by the step previewer when writing new tests in a feature file.

**Reply by gasparnagy** [reply]
@RikvanSpreuwel Ah sorry. Actually that is not about *detection* (they are detected correctly as regex), but about visualization. [This class](https://github.com/reqnroll/Reqnroll.VisualStudio/blob/main/Reqnroll.VisualStudio/Editor/Completions/StepDefinitionSampler.cs) does it. You can try to improve it for regex step defs and send a PR, but there is no built-in option to make what you would like to achieve.

**Reply by RikvanSpreuwel** [reply]
@gasparnagy Thank you, i'll try to find time to fix this. 

If i understand correctly, the problem is that the `RegexStepDefinitionsExpressionAnalyzer` thinks the expression contains only simple text?
Would it be possible to use the `ICucumberExpressionDetector` here (that can be overwritten with the ForceRegexPlugin) to check if it is not a cucumber expression, and if so, simply return the `regexTextCore`?

**Reply by gasparnagy** [reply]
@RikvanSpreuwel I'm not sure if that is the problem, but maybe. 

The step definitions are processed using the compiled test project, so the `ICucumberExpressionDetector` overrides should be also applied. I'm afraid, you need to debug it through to be able to figure out what is wrong, really.

**Reply by RikvanSpreuwel** [reply]
@gasparnagy Allright, thanks for the information. I will debug it at some later time to figure out what is wrong and how to fix it

**Reply by gasparnagy** [reply]
@RikvanSpreuwel Sure. Please let me know if you run into any troubles.

**Reply by RikvanSpreuwel** [reply]
@gasparnagy The problem was indeed not necessarily that steps were not detected as regex. Instead, I think the problem was that a set of options (defined like '(option1|option2|option3)') is converted to show the parameter type ('[string]').

As a fix, I have added that after separating the expression into parts, I check for parameter parts if it is a list of options in this format. If so, I do not convert it to the parameter type but add the part instead.
See PR: https://github.com/reqnroll/Reqnroll.VisualStudio/pull/55
